File: design_pattern/solid/dip/solid_dip.dart
enum Relationship { parent, child, sibling }

class Person {
  final String name;
  Person(this.name);
}

class Info {
  final Person from;
  final Relationship relationship;
  final Person to;

  Info(this.from, this.relationship, this.to);
}

// 低レベルモジュールの抽象化
abstract interface class RelationshipBrowser {
  List<Person> findAllChildrenOf(String name);
}

// 低レベルモジュールの実装
class Relationships implements RelationshipBrowser {
  final List<Info> _relations = [];

  List<Info> get relations => List.unmodifiable(_relations);

  void addParentAndChild(Person parent, Person child) {
    _relations
      ..add(Info(parent, Relationship.parent, child))
      ..add(Info(child, Relationship.child, parent));
  }

  @override
  List<Person> findAllChildrenOf(String name) => _relations
      .where(
          (r) => r.from.name == name && r.relationship == Relationship.parent)
      .map((r) => r.to)
      .toList();
}

// 高レベルモジュール
class Research {
  const Research._();

  static void findChildren(RelationshipBrowser browser, String name) {
    final children = browser.findAllChildrenOf(name);
    for (final child in children) {
      print('$nameは${child.name}の親です');
    }
  }
}

void main() {
  final parent = Person('John');
  final child1 = Person('Chris');
  final child2 = Person('Matt');

  final relationships = Relationships();
  relationships.addParentAndChild(parent, child1);
  relationships.addParentAndChild(parent, child2);

  Research.findChildren(relationships, 'John');
}

File: design_pattern/solid/dip/solid_dip.go
package main

/// DIP (Dependency Inversion Principle)
/// 1. 上位モジュールは下位レベルのモジュールに依存してはならない
/// 2. 両方とも抽象に依存すべき
/// 3. 抽象は実装に依存すべきではない
/// 4. 実装が抽象に依存すべき
///
/// Goの場合、抽象クラスや基底クラスはインターフェースを指す

import "fmt"

type Relationship int

const (
	Parent Relationship = iota
	Child
	Sibling
)

type Person struct {
	name string
}

type Info struct {
	from         *Person
	relationship Relationship
	to           *Person
}

// low-level module
type RelationshipBrowser interface {
	FindAllChildrenOf(name string) []*Person
}

// low-level module
type Relationships struct {
	relations []Info
}

func (rs *Relationships) FindAllChildrenOf(name string) []*Person {
	result := make([]*Person, 0)

	for i, v := range rs.relations {
		if v.relationship == Parent && v.from.name == name {
			result = append(result, rs.relations[i].to)
		}
	}

	return result
}

func (rs *Relationships) AddParentAndChild(parent, child *Person) {
	rs.relations = append(rs.relations, Info{parent, Parent, child})
	rs.relations = append(rs.relations, Info{child, Child, parent})
}

// high-level module
type Research struct {
	// DIP違反: low-level moduleに依存している
	// relationships Relationships
	browser RelationshipBrowser
}

func (r *Research) Investigate(parent string) {
	// DIP違反： researchモジュールがrelationshipsモジュールの内部構造に依存している
	//relations := r.relationships.relations
	//for _, rel := range relations {
	//	if rel.from.name == "John" &&
	//		rel.relationship == Parent {
	//		fmt.Println("John has a child called", rel.to.name)
	//	}
	//}

	for _, p := range r.browser.FindAllChildrenOf(parent) {
		fmt.Printf("%s has a child called %s\n", parent, p.name)
	}
}

func main() {
	parent := Person{"John"}
	child1 := Person{"Chris"}
	child2 := Person{"Matt"}

	// low-level module
	relationships := Relationships{}
	relationships.AddParentAndChild(&parent, &child1)
	relationships.AddParentAndChild(&parent, &child2)

	research := Research{&relationships}
	research.Investigate("John")
}
